### ISSUE TITLE
duplicated tree-sitter parsers in runtimepath

### ISSUE THREAD

**Issue by boltlessengineer** [nvim-neorocks/rocks.nvim#354]
I’m not sure if here is right place for this issue, sorry if I’m wrong.

I’ve installed tree-sitter parsers via `:Rocks install tree-sitter-javscript`, and because I’m using aarch64 machine which NURR doesn’t support yet, luarocks tried to build parser locally. Installation is fine but now I can see three duplicated parsers from runtimepath.

```
vim.treesitter: require("vim.treesitter.health").check()

- Nvim runtime ABI version: 14
- OK Parser: rust                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-rust/parser/rust.so
- OK Parser: javascript           ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-javascript/parser/javascript.so
- OK Parser: http                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-http/parser/http.so
- OK Parser: html                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-html/parser/html.so
- OK Parser: css                  ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-css/parser/css.so
- OK Parser: toml                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/site/pack/luarocks/opt/tree-sitter-toml/parser/toml.so
- OK Parser: bash                 ABI: 14, path: /usr/local/lib/nvim/parser/bash.so
- OK Parser: c                    ABI: 14, path: /usr/local/lib/nvim/parser/c.so
- OK Parser: lua                  ABI: 14, path: /usr/local/lib/nvim/parser/lua.so
- OK Parser: markdown             ABI: 14, path: /usr/local/lib/nvim/parser/markdown.so
- OK Parser: markdown_inline      ABI: 14, path: /usr/local/lib/nvim/parser/markdown_inline.so
- OK Parser: python               ABI: 14, path: /usr/local/lib/nvim/parser/python.so
- OK Parser: query                ABI: 14, path: /usr/local/lib/nvim/parser/query.so
- OK Parser: vim                  ABI: 14, path: /usr/local/lib/nvim/parser/vim.so
- OK Parser: vimdoc               ABI: 14, path: /usr/local/lib/nvim/parser/vimdoc.so
- OK Parser: css                  ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-css/scm-1/parser/css.so
- OK Parser: html                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-html/scm-1/parser/html.so
- OK Parser: http                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-http/scm-1/parser/http.so
- OK Parser: javascript           ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-javascript/scm-1/parser/javascript.so
- OK Parser: rust                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-rust/scm-1/parser/rust.so
- OK Parser: toml                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-toml/scm-1/parser/toml.so
- OK Parser: css                  ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/css.so
- OK Parser: html                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/html.so
- OK Parser: http                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/http.so
- OK Parser: javascript           ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/javascript.so
- OK Parser: rust                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/rust.so
- OK Parser: toml                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser/toml.so
```

![IMG_0938](https://github.com/nvim-neorocks/luarocks-build-treesitter-parser/assets/60088301/df04dd5a-1f43-4360-b6b8-546f266c2d64)

I even get duplicated captures from tree-sitter attached buffer.

Can we prevent luarocks doing this?

**Comment by boltlessengineer** [comment]
I can’t see any issues beside `:Inspect` and `:checkhealth vim.treesitter` yet. But it really feels weird to see multiple captures there.

**Comment by mrcjkb** [comment]
@boltlessengineer this should be fixed in version `2.29.1`.
As it doesn't cause any actual problems, I haven't added logic for removing the `parser` symlink in the `rocks_rtp` directory (In your case, this is `/home/ubuntu/.local/share/nvim_rocks/rocks/rocks_rtp/parser`).
So you'll have to remove it manually after updating.

**Comment by boltlessengineer** [comment]
I’m 100% fine with that. We should not put that as an overhead.

**Comment by mrcjkb** [comment]
This output of your health check is suspicious:

>```
>- OK Parser: css                  ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-css/scm-1/parser/css.so
>- OK Parser: html                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-html/scm-1/parser/html.so
>- OK Parser: http                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-http/scm-1/parser/http.so
>- OK Parser: javascript           ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-javascript/scm-1/parser/javascript.so
>- OK Parser: rust                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-rust/scm-1/parser/rust.so
>- OK Parser: toml                 ABI: 14, path: /home/ubuntu/.local/share/nvim_rocks/rocks/lib/luarocks/rocks-5.1/tree-sitter-toml/scm-1/parser/toml.so
>```

The paths are on the `package.cpath` and shouldn't normally be on Neovim's runtimepath.

If you're seeing duplicate queries, that probably means you have something on the runtimepath that isn't added by rocks.nvim.

In my tests (with tree-sitter-toml), I don't see duplicate captures with `:Inspect`.

**Comment by mrcjkb** [comment]
> I can still see pairs of identical parsers form health check

It looks like your Neovim install is syncing the `package.cpath` with the `runtimepath`, which it doesn't always do, afaik (unless this has been fixed). But that's fine. They're the same files. Neovim will use the first parser it finds.

> so I guess highlight.scm query for ecma or javascript is duplicated somewhere

Most query files have multiple different queries for the same capture group. Sometimes, more than one query can match. In this case, `:Inspect` will show capture groups from all matches and Neovim will apply the highlight for the last capture.

> When I run :TSEditQuery highlights after installing nvim-treesitter, I get 3 duplicated highlight.scm query paths in option, so it seems to be duplicated query files are the problem.

If you use nvim-treesitter (9.2 or the `master` branch) and parsers from luarocks, you will get duplicate queries. That's because those versions of nvim-treesitter still ship with queries on the rtp.
You can solve this by switching to the nvim-treesitter `main` branch.

If you need nvim-treesitter in order to support legacy plugins that still use it as a dependency, I have packaged a [`nvim-treesitter-legacy-api`](https://luarocks.org/modules/neorocks/nvim-treesitter-legacy-api) rock. 

See also: https://github.com/nvim-neorocks/rocks.nvim/discussions/325

The paths in your output are all from the nvim-treesitter plugin. So I doubt that has anything to do with rocks.nvim.

**Comment by boltlessengineer** [comment]
> It looks like your Neovim install is syncing the package.cpath with the runtimepath, which it doesn't always do, afaik (unless this has been fixed). But that's fine. They're the same files. Neovim will use the first parser it finds.

You were right. That was the problem. My neovim’s version was `0.10.0` and upgrading to `0.10.1` fixed the issue.
Thank you so much for your active help!

**Comment by mrcjkb** [comment]
> My neovim’s version was 0.10.0 and upgrading to 0.10.1 fixed the issue.

There's a neovim 0.10.1?

Now I'm confused - I don't know if this is a bug that has been fixed in 0.10.0 and reintroduced, or if it's a bug that was introduced in 0.10.0 and has now been fixed.
